net/url/url.h

```diff
@@ -102,43 +106,157 @@ struct URL {
       path = "/";
     }
 
-    if (protocol.empty()) {
-      if (!previous_protocol.empty()) {
-        protocol = previous_protocol;
+    if (scheme.empty()) {
+      if (!previous_scheme.empty()) {
+        scheme = previous_scheme;
       } else {
-        protocol = DefaultProtocolForPort(port);
+        scheme = DefaultSchemeForPort(port);
       }
     }
 
     if (port == 0) {
-      port = DefaultPortForProtocol(protocol);
+      port = DefaultPortForScheme(scheme);
     }
   }
 
-  URL(const std::string& url, const URL& previous)
-      : URL(url, previous.protocol, previous.host, previous.port) {}
+  URLWithoutParameters(const std::string& url, const URLWithoutParameters& previous)
+      : URLWithoutParameters(url, previous.scheme, previous.host, previous.port) {}
 
   std::string ComposeURL() const {
-    std::ostringstream os;
-    if (!protocol.empty()) {
-      os << protocol << "://";
+    if (!host.empty()) {
+      std::ostringstream os;
+      if (!scheme.empty()) {
+        os << scheme << "://";
+      }
+      os << host;
+      if (port != DefaultPortForScheme(scheme)) {
+        os << ':' << port;
+      }
+      os << path;
+      return os.str();
+    } else {
+      // If no host is specified, it's just the path: No need to put scheme and port.
+      return path;
+    }
+  }
+
+  static int DefaultPortForScheme(const std::string& scheme) {
+    // We don't really "support" other schemes yet -- D.K.
+    if (scheme == "http") {
+      return 80;
+    } else if (scheme == "https") {
+      return 443;
+    } else {
+      return 0;
+    }
+  }
+
+  static std::string DefaultSchemeForPort(int port) { return port == 80 ? "http" : ""; }
+};
+
+struct URLParameters {
+  URLParameters() = default;
+  URLParameters(std::string url) {
+    const size_t pound_sign_index = url.find('#');
+    if (pound_sign_index != std::string::npos) {
+      fragment = url.substr(pound_sign_index + 1);
+      url = url.substr(0, pound_sign_index);
+    }
+    const size_t question_mark_index = url.find('?');
+    if (question_mark_index != std::string::npos) {
+      parameters_vector = strings::SplitIntoKeyValuePairs(url.substr(question_mark_index + 1), '=', '&');
+      for (auto& it : parameters_vector) {
+        it.second = DecodeURIComponent(it.second);
```

Comment:
The key could also be encoded, like `foo%20bar=value` should give a `"foo bar"` key.